src/huggingface_hub/repocard.py

```diff
@@ -99,3 +104,153 @@ def metadata_eval_result(
+        filename="README.md",
+        repo_type=repo_type,
+        use_auth_token=token,
+        force_download=True,
+    )
+    existing_metadata = metadata_load(filepath)
+
+    for key in metadata:
+        # update all fields except model index
+        if key != "model-index":
+            if key in existing_metadata and not overwrite:
+                if existing_metadata[key] != metadata[key]:
+                    raise ValueError(
+                        f"""You passed a new value for the existing meta data field '{key}'. Set `overwrite=True` to overwrite existing metadata."""
+                    )
+            else:
+                existing_metadata[key] = metadata[key]
+        # update model index containing the evaluation results
+        else:
+            if "model-index" not in existing_metadata:
+                existing_metadata["model-index"] = metadata["model-index"]
+            else:
+                # the model-index contains a list of results as used by PwC but only has one element thus we take the first one
+                existing_metadata["model-index"][0][
+                    "results"
+                ] = _update_metadata_model_index(
+                    existing_metadata["model-index"][0]["results"],
+                    metadata["model-index"][0]["results"],
+                    overwrite=overwrite,
+                )
+
+    # save and push to hub
+    metadata_save(filepath, existing_metadata)
+
+    HfApi().upload_file(
+        path_or_fileobj=filepath,
+        path_in_repo="README.md",
+        repo_id=repo_id,
+        repo_type=repo_type,
+        identical_ok=True,
+        token=token,
+    )
+
+
+def _update_metadata_model_index(existing_results, new_results, overwrite=False):
+    """
+    Updates the model-index fields in the metadata. If results with same unique
+    features exist they are updated, else a new result is appended. Updating existing
+    values is only possible if `overwrite=True`.
+
+    Args:
+        new_metrics (`list`):
+            List of new metadata results
+        existing_metrics (`list`):
+            List of existing metadata results
+        overwrite (`bool`, *optional*, defaults to `False`):
+            If set to `True` an existing metric values can be overwritten, otherwise
+            attempting to overwrite an existing field will cause an error.
+
+    Returns:
+        `list`: List of updated metadata results
+    """
+    for new_result in new_results:
+        try:
+            existing_result = next(
+                x
+                for x in existing_results
+                if all(x[feat] == new_result[feat] for feat in UNIQUE_RESULT_FEATURES)
+            )
+            existing_result["metrics"] = _update_metadata_results_metric(
+                new_result["metrics"],
+                existing_result["metrics"],
+                overwrite=overwrite,
+            )
+        except StopIteration:
+            existing_results.append(new_result)
```

Comment:
Is this really needed? I think try/except to do this is a bit less readable and could be more prone to introduce bugs. Maybe the for loop could be more explicit. I think that was how it was before, I personally find that more readable and easier to maintain.

Reply:
I have no strong opinion here - happy to revert. @adrinjalali?

Reply:
The for loop is slower, and try/except clauses are quite pythonic. I do think the current state is quite better than a for loop. One can always add a bit of comment on when the exception is raised if that helps with readability.

Reply:
I don't think speed is important here as we iterate through a dictionary with a handful of entries while pulling/pushing to the hub in the same function which is probably 1000x slower.

Reply:
I do find the for/loop implementation less readable. Removing the outer loop would make it quite complicated and not very readable, that's why I didn't suggest to remove it.

As for consistency, I don't think it's about staying consistent with the rest of the code. Sometimes a for loop makes sense, sometimes it doesn't.  And as a team when we work on a codebase, sometimes we find better ways to do things, and that's fine, and I don't think we should hold back because we have done things in a certain way so far. To me it's okay to look at a codebase and notice old stuff vs new stuff. The way people do things changes and new code can look different than the old code and that's fine to me.

But if both of you think the for loop is more readable than this, then sure, change it. To me it's the other way around.